Expose early stopping hyper-parameters for all tasks

supersoob opened the pull request:
This PR includes

- Add early stopping parameters in config.py 
- Overwrite early stopping hook
- hyper-parameter change for multi-class classification to enable early stopping

Commit:
Early stop as hp

Commit:
Early stop metric added

Commit:
override early stopping hook

Commit:
inter_patience

Commit:
fix merge

Commit:
spaces

Commit:
Patience ui header

Commit:
Merge branch 'releases/v0.3.0-sc1.2' of https://github.com/openvinotoolkit/training_extensions into early_stop_hp

Commit:
test

Commit:
Fix seg

Commit:
warning for patience

Commit:
adaptive remove

Commit:
fix bug when early stop is false

Commit:
warning sign in patience

Commit:
Merge branch 'releases/v0.3.0-sc1.2' of https://github.com/openvinotoolkit/training_extensions into early_stop_hp

supersoob commented:
run ote-test

JihwanEom commented on external/model-preparation-algorithm/configs/classification/configuration.yaml:
```diff
@@ -123,15 +123,32 @@ learning_parameters:
+  patience:
+    affects_outcome_of: TRAINING
+    default_value: 10
+    description: Number of epochs without improvement before the system will automatically stop training.
```
How about "Training will be stopped if the model is not improved during patience."

JihwanEom commented on external/model-preparation-algorithm/configs/detection/configuration.yaml:
```diff
@@ -119,6 +119,37 @@ learning_parameters:
+  enable_early_stopping:
+    affects_outcome_of: TRAINING
+    default_value: true
+    description: Adaptive early exit from training when accuracy isn't changed or decreased for several epochs.
```
I'm confused with "Adaptive".. Does that mean early stopping by adaptive interval?

JihwanEom commented on external/model-preparation-algorithm/mpa_tasks/apis/classification/task.py:
```diff
@@ -381,10 +375,13 @@ def _patch_evaluation(self, config: MPAConfig):
+            config.early_stop_metric = 'mAP'
         elif self._hierarchical:
             cfg.metric = ['MHAcc', 'avgClsAcc', 'mAP']
+            config.early_stop_metric = 'mAP'
```
Could you change into `MHAcc` or `accuracy`? mAP can be always 0.0 if hierarchical project is consisted only with single-label classification groups.
For 'accuracy', please refer this PR : https://github.com/openvinotoolkit/training_extensions/pull/1215

JihwanEom commented on external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py:
```diff
@@ -330,9 +332,7 @@ def _patch_evaluation(config: MPAConfig):
-        for cfg in config.get('custom_hooks', []):
-            if 'EarlyStoppingHook' in cfg.type:
-                cfg.metric = 'mAP'
+        config.early_stop_metric = 'mAP'
```
@jaegukhyun , I can't remember exactly, does detection not use 'bbox_mAP'?

jaegukhyun replied:
'mAP' is right

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -155,10 +156,48 @@ def _initialize(self, dataset=None, output_model=None, export=False):
                 self._recipe_cfg.merge_from_dict(recipe_hparams)
+
+            # Put early stop hook
+            if self._recipe_cfg.early_stop is True:
```
I think it would be good to subtract the code related to early stop separately as a function.

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -155,10 +156,48 @@ def _initialize(self, dataset=None, output_model=None, export=False):
+                early_stop_hook['patience'] = patience
+                early_stop_hook['metric'] = metric
+
+                self.update_override_configurations({"custom_hooks" : [early_stop_hook]})
```
we can use `update_or_add_custom_hook`

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -155,10 +156,48 @@ def _initialize(self, dataset=None, output_model=None, export=False):
+
+                self.update_override_configurations({"custom_hooks" : [early_stop_hook]})
+            else:
+                custom_hooks = self._recipe_cfg.get('custom_hooks',[])
```
I think this can be solved simply by increasing the start value of early-stop to more than max_epoch. What do you think?

supersoob replied:
Thank you for suggestion, but it kind of looks like a hack. Completely removing the early stopping hook can also prevent from redundant part such as compare between current epoch and start epoch for every epoch in early stopping hook.

Commit:
Adaptive validation interval added

Commit:
Merge branch 'releases/v0.3.0-sc1.2' of https://github.com/openvinotoolkit/training_extensions into early_stop_hp

Commit:
iseg

Commit:
Refator

supersoob commented:
Hi, I refactored the codes reflecting your reviews. I changed the descriptions for early stopping hp and added new hp for adaptive interval. And I also wanted to make a function for early stopping update/removal, but it seems like there's no where to put related functions. Instead, I downsized the codes for early stopping hp. Please take a look again.

Commit:
Fix minor errors

Commit:
params update in guide

Commit:
Typo

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/config.py:
```diff
@@ -77,6 +77,44 @@ class BaseLearningParameters(ParameterGroup):
+            affects_outcome_of=ModelLifecycle.TRAINING
+        )
+        
+        patience = configurable_integer(
```
I think this is too general a variable name. This is the patience of early stopping, but it seems to be used together elsewhere.

supersoob replied:
Do you suggest to change the variable name? Like early_stop_patience/early_stop_iteration_patience?

harimkang replied:
Yes, I think that is appropriate.

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/config.py:
```diff
@@ -77,6 +77,44 @@ class BaseLearningParameters(ParameterGroup):
+            affects_outcome_of=ModelLifecycle.TRAINING
+        )
+
+        iteration_patience = configurable_integer(
```
This variable also looks too general.

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py:
```diff
@@ -330,9 +339,7 @@ def _patch_evaluation(config: MPAConfig):
-        for cfg in config.get('custom_hooks', []):
-            if 'EarlyStoppingHook' in cfg.type:
-                cfg.metric = 'mAP'
+        config.early_stop_metric = 'mAP'
```
To always be equal to the metric of evaluation, looks better to use early_stop_metric = cfg.metric. (The same goes for other tasks)

supersoob replied:
I put the early stopping check code in task.py but cls uses mutiple metrics, so early_stop_metric = cfg.metric will bring the lists of metrics in task.py in apis folder, uncertain what to choose for early stopping metric.

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py:
```diff
@@ -149,9 +149,17 @@ def _init_recipe_hparam(self) -> dict:
         lr_config = ConfigDict(warmup_iters=warmup_iters) if warmup_iters > 0 \
             else ConfigDict(warmup_iters=warmup_iters, warmup=None)
+        
+        if self._hyperparams.learning_parameters.enable_early_stopping is True:
```
It looks good to remove `is True`.

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -181,6 +184,34 @@ def _initialize(self, dataset=None, output_model=None, export=False):
+
+        # Add/remove early stop hook
+        if 'early_stop' in self._recipe_cfg:
+            remove_custom_hook(self._recipe_cfg, 'EarlyStoppingHook')
```
Does this mean that this will not use the regular 'EarlyStoppingHook' but only the 'LazyEarlyStoppingHook'?

supersoob replied:
Yes, you are right.

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -181,6 +184,34 @@ def _initialize(self, dataset=None, output_model=None, export=False):
+            remove_custom_hook(self._recipe_cfg, 'EarlyStoppingHook')
+            early_stop = self._recipe_cfg.get('early_stop', False)
+            logger.info(f"early_stop : {early_stop}")
+            if early_stop is not False:
```
`if early_stop` looks appropriate

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -181,6 +184,34 @@ def _initialize(self, dataset=None, output_model=None, export=False):
 
+        # Add/remove adaptive interval hook
+        if 'use_adaptive_interval' in self._recipe_cfg:
+            if self._recipe_cfg.use_adaptive_interval is True:
```
I don't think we need this if statement.

supersoob replied:
What statement are you talking? You reviewed the same part as below.

supersoob replied:
I can't remove that if statement. Otherwise, it will cause error in the export stage. initialize function where init_recipe_hparam is initiated is called only when it's not export.

harimkang replied:
What I mean is that if statement right above and this if statement can be combined into one, and I don't think it's going to cause an error because of this during export, so what do you think?

supersoob replied:
Your suggestion is really good, but it raises error in else for cls/seg tasks -> self._recipe_cfg.pop('adaptive_validation_interval')

use_adaptive_interval is the parameter only for detection.(because it did not work in other tasks.. must test at other time) So it needs to check first if use_adaptive_interval is actually in the recipe_cfg. and then need to pop and remove adaptive_validation_interval config only for det task when use_adaptive_interval is set to False.

harimkang replied:
I think we can prevent errors by using `self._recipe_cfg.pop('adaptive_validation_interval', None)`.

supersoob replied:
I modified this part. Please check it again!

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -181,6 +184,34 @@ def _initialize(self, dataset=None, output_model=None, export=False):
         self._precision = [ModelPrecision.FP32]
 
+        # Add/remove adaptive interval hook
+        if 'use_adaptive_interval' in self._recipe_cfg:
```
I think `self._recipe_cfg.get('use_adaptive_interval', False)` is better

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -181,6 +184,34 @@ def _initialize(self, dataset=None, output_model=None, export=False):
+            if self._recipe_cfg.use_adaptive_interval is True:
+                ada_interval_cfg = self._recipe_cfg.get('adaptive_validation_interval', None)
+                if ada_interval_cfg is None:
+                    self._recipe_cfg.adaptive_validation_interval = dict(max_interval=5)
```
Doesn't this cause an error if `adaptive_validation_interval `does not exist?
and If it's not an issue that occurs, it seems that `self._recipe_cfg.adaptive_validation_interval = self._recipe_cfg.get('adaptive_validation_interval', dict(max_interval=5))` is more appropriate.

supersoob replied:
Doesn't it just overwrite to its own value? Can you explain why it's more appropriate? And why do you think it causes error? I'm just curious.

harimkang replied:
I was just thinking about maybe we could reduce the code to three lines -> 1 line code. Also I thought there would be an issue if `self._recipe_cfg` didn't have `adaptive_validation_interval`.

supersoob replied:
I modified this part. Please check it again!

Commit:
Refactor and change hp name

Commit:
Modify codes for ada_itv_hook

Commit:
Typo

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py:
```diff
@@ -149,9 +149,17 @@ def _init_recipe_hparam(self) -> dict:
         return ConfigDict(
             optimizer=ConfigDict(lr=self._hyperparams.learning_parameters.learning_rate),
             lr_config=lr_config,
+            early_stop=early_stop,
```
Of course, the adaptive_interval of the segmentation, the default is still set to False, but can't we enable this function?

harimkang commented on external/model-preparation-algorithm/mpa_tasks/apis/task.py:
```diff
@@ -181,6 +184,30 @@ def _initialize(self, dataset=None, output_model=None, export=False):
 
+        # Add/remove adaptive interval hook
+        if self._recipe_cfg.get('use_adaptive_interval', False):
+            self._recipe_cfg.adaptive_validation_interval = self._recipe_cfg.get('adaptive_validation_interval', dict(max_interval=5))
```
`self._recipe_cfg` is `ConfigDict`. Therefore, `self._recipe_cfg.adaptive_validation_interval` will raise the `AttributeError` if the key `adaptive_validation_interval` does not exist.

supersoob replied:
Sorry, I don't understand. As far as I know, ConfigDict can insert keys and values directly by using `self._recipe_cfg.adaptive_validation_interval`. I tested it works.

harimkang replied:
I guess I got it wrong. I'm sorry.

harimkang reviewed (dismissed):
If the adaptive interval of detection works well in SC, I think it can be merged.

harimkang commented:
+Currently, the minimum value of patience is 0 (editable in UI), so I think we need to check if it works well.

supersoob commented:
Test results
![image](https://user-images.githubusercontent.com/37075645/188789280-da010be4-2a3d-4560-b411-6c2b6f4c2cb6.png)

Commit:
Fix typo in warning

supersoob commented:
I checked all tasks are working well with early stopping hp. Also working well when patience=0. I will update the hyperparameters resulted from early stop exp. for classification soon. Please review the related [MPA PR](https://github.com/openvinotoolkit/model_preparation_algorithm/pull/56), too

Commit:
default hp update

Commit:
hp update

eunwoosh committed:
update hpo_config file based on updated classification default hyper parameters

JihwanEom reviewed (commented):
Could you resolve below issue?
<img width="815" alt="image" src="https://user-images.githubusercontent.com/72460430/190957645-414085f2-b8ca-4191-bbbe-192551794dbc.png">

https://app.codacy.com/gh/openvinotoolkit/training_extensions/pullRequest?prid=10125167

Commit:
Codacy Static Code Analysis

Commit:
QUICK_START_GUIDE.md parameters

supersoob commented:
> Could you resolve below issue? <img alt="image" width="815" src="https://user-images.githubusercontent.com/72460430/190957645-414085f2-b8ca-4191-bbbe-192551794dbc.png">
> 
> https://app.codacy.com/gh/openvinotoolkit/training_extensions/pullRequest?prid=10125167

I think it doesn't matter if it fails that check. There are longer lines in QUICK_START_GUIDE.md. Old PRs are merged even though it fails some codacy code analysis.

harimkang reviewed (commented):
There are no other comments on the implementation of the code. 
There are some modified model template values and hpo_config values, what kind of background does this have? (Maybe I missed some of the experimental results...)

(I missed it briefly, but hpo value is modified by Eun-Woo, Forget about hpo_config.)

Commit:
Change MPA submodule

Commit:
Make start epoch param for early stopping

Commit:
hp for inst-seg

Commit:
Change MPA submodule commit
